SonarQube.Client/RoslynExporterAdapter/NuGetPackageInfoGenerator.cs

```diff
@@ -0,0 +1,59 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using SonarQube.Client.Messages;
+using SonarQube.Client.Models;
+
+namespace SonarQube.Client.RoslynExporterAdapter
+{
+    public static class NuGetPackageInfoGenerator
+    {
+        public static IEnumerable<NuGetPackageInfoResponse> GetNuGetPackageInfos(IEnumerable<SonarQubeRule> activeRules,
+            IDictionary<string, string> sonarProperties)
+        {
+            var propertyPrefixes = GetPluginPropertyPrefixes(activeRules);
+            var packages = new List<NuGetPackageInfoResponse>();
+
+            foreach (var partialRepoKey in propertyPrefixes)
+            {
+                if (!sonarProperties.TryGetValue($"{partialRepoKey}.analyzerId", out var analyzerId) ||
+                    !sonarProperties.TryGetValue($"{partialRepoKey}.pluginVersion", out var pluginVersion))
+                {
+                    continue;
+                }
+
+                packages.Add(new NuGetPackageInfoResponse { Id = analyzerId, Version = pluginVersion });
+            }
+
+            return packages;
+        }
+
+        private static IEnumerable<string> GetPluginPropertyPrefixes(IEnumerable<SonarQubeRule> rules)
+        {
+            var partialRepoKeys = new HashSet<string>();
+            partialRepoKeys.UnionWith(rules.Select(r => r.TryGetRoslynPluginPropertyPrefix()));
+            partialRepoKeys.Remove(null);
```

Comment:
The null is a bit confusing.
Also, maybe instead of a hashset, we can use Distinct(), and then Where(x=> !string.IsNullOrEmpty(x)

Reply:
Ok, if you find it confusing,